File: src/utils/observer.ts
/**@noSelf */
export interface IObservable<T> {
    subscribe(fn: (data: T) => void): void;
    unsubscribe(fn: (data: T) => void): void;
    get(): T;
    set(value: T): void;
}

/**@noSelf */
export interface IObservableNumberVal extends IObservable<number> {
    sub(val: number): void;
    add(val: number): void;
}

const createObservableVal = <T>(value: T, receive_msg_when_subscribe = false): IObservable<T> => {
    let observers: Array<(data: T) => void> = [];
    let value_ = value;

    const subscribe = (fn: (data: T) => void) => {
        observers.push(fn);
        if (receive_msg_when_subscribe) {
            fn(value_);
        }
    };

    const unsubscribe = (fn: (data: T) => void) => {
        observers = observers.filter((subscriber) => subscriber !== fn);
    };

    const setValue = (data: T) => {
        value_ = data;
        observers.forEach((subscriber) => subscriber(value_));
    };

    const getValue = () => {
        return value_;
    };

    return {
        get: getValue,
        set: setValue,
        subscribe,
        unsubscribe,
    };
};

const createObservableNumberVal = (value: number, receive_msg_when_subscribe = false): IObservableNumberVal => {
    const observableVal = createObservableVal(value, receive_msg_when_subscribe);

    const add = (v: number) => {
        observableVal.set(observableVal.get() + v);
    };

    const sub = (v: number) => {
        observableVal.set(observableVal.get() - v);
    };

    return {
        ...observableVal,
        add,
        sub,
    };
};

const createObservableNumberValWithPre = (value: number, receive_msg_when_subscribe = false) => {

    let observers: Array<(new_value: number, pre_value: number) => void> = [];
    let value_ = value;
    let pre_value = value;
    const subscribe = (fn: (new_value: number, pre_value: number) => void) => {
        observers.push(fn);
        if (receive_msg_when_subscribe) {
            fn(value_, pre_value);
        }
    };

    const unsubscribe = (fn: (new_value: number, pre_value: number) => void) => {
        observers = observers.filter((subscriber) => subscriber !== fn);
    };

    const setValue = (data: number) => {
        pre_value = value_;
        value_ = data;
        observers.forEach((subscriber) => subscriber(value_, pre_value));
    };

    const getValue = () => {
        return value_;
    };

    return {
        get: getValue,
        set: setValue,
        subscribe,
        unsubscribe,
    };
};

export const ObservableVal = createObservableVal;
export const ObservableNumberVal = createObservableNumberVal;

export const ObservableNumberValWithPre = createObservableNumberValWithPre;

File: app/utils/observer.lua
local ____lualib = require("lualib_bundle")
local __TS__ArrayFilter = ____lualib.__TS__ArrayFilter
local __TS__ArrayForEach = ____lualib.__TS__ArrayForEach
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local ____exports = {}
local function createObservableVal(value, receive_msg_when_subscribe)
    if receive_msg_when_subscribe == nil then
        receive_msg_when_subscribe = false
    end
    local observers = {}
    local value_ = value
    local function subscribe(fn)
        observers[#observers + 1] = fn
        if receive_msg_when_subscribe then
            fn(value_)
        end
    end
    local function unsubscribe(fn)
        observers = __TS__ArrayFilter(
            observers,
            function(____, subscriber) return subscriber ~= fn end
        )
    end
    local function setValue(data)
        value_ = data
        __TS__ArrayForEach(
            observers,
            function(____, subscriber) return subscriber(value_) end
        )
    end
    local function getValue()
        return value_
    end
    return {get = getValue, set = setValue, subscribe = subscribe, unsubscribe = unsubscribe}
end
local function createObservableNumberVal(value, receive_msg_when_subscribe)
    if receive_msg_when_subscribe == nil then
        receive_msg_when_subscribe = false
    end
    local observableVal = createObservableVal(value, receive_msg_when_subscribe)
    local function add(v)
        observableVal.set(observableVal.get() + v)
    end
    local function sub(v)
        observableVal.set(observableVal.get() - v)
    end
    return __TS__ObjectAssign({}, observableVal, {add = add, sub = sub})
end
local function createObservableNumberValWithPre(value, receive_msg_when_subscribe)
    if receive_msg_when_subscribe == nil then
        receive_msg_when_subscribe = false
    end
    local observers = {}
    local value_ = value
    local pre_value = value
    local function subscribe(fn)
        observers[#observers + 1] = fn
        if receive_msg_when_subscribe then
            fn(value_, pre_value)
        end
    end
    local function unsubscribe(fn)
        observers = __TS__ArrayFilter(
            observers,
            function(____, subscriber) return subscriber ~= fn end
        )
    end
    local function setValue(data)
        pre_value = value_
        value_ = data
        __TS__ArrayForEach(
            observers,
            function(____, subscriber) return subscriber(value_, pre_value) end
        )
    end
    local function getValue()
        return value_
    end
    return {get = getValue, set = setValue, subscribe = subscribe, unsubscribe = unsubscribe}
end
____exports.ObservableVal = createObservableVal
____exports.ObservableNumberVal = createObservableNumberVal
____exports.ObservableNumberValWithPre = createObservableNumberValWithPre
return ____exports
